linkerd/distribute/src/keys.rs

```diff
@@ -0,0 +1,160 @@
+use rand::{
+    distributions::{WeightedError, WeightedIndex},
+    prelude::Distribution as _,
+    Rng,
+};
+use std::{collections::HashMap, hash::Hash};
+
+/// Uniquely identifies a key/backend pair for a distribution. This allows
+/// backends to have the same key and still participate in request distribution.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct KeyId {
+    idx: usize,
+}
+
+impl KeyId {
+    pub(crate) fn new(idx: usize) -> Self {
+        Self { idx }
+    }
+}
+
+#[derive(Debug)]
+pub struct ServiceKeys<K> {
+    ids: Vec<KeyId>,
+    keys: HashMap<KeyId, K>,
```

Comment:
Note that this crate already uses an AHashMap (faster hash algo), so it would be suitable to use that here as well
```suggestion
    keys: ahash::AHashMap<KeyId, K>,
```

Reply:
Done